rust/crypto/src/ghash.rs

```diff
@@ -0,0 +1,78 @@
+//
+// Copyright 2021 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use crate::polyval::Polyval;
+use crate::{Error, Result};
+use std::convert::TryInto;
+
+pub struct Ghash {
+    polyval: Polyval,
+}
+
+fn reverse_bytes_16(c: &[u8]) -> [u8; 16] {
+    assert!(c.len() == 16);
+    u128::from_be_bytes(c.try_into().expect("Verified size")).to_le_bytes()
```

Comment:
Ugh, I'd consider it a Rust optimizer bug that this is faster than the generic version below. Oh well.

Reply:
Yeah I was kind of sad but the difference is stark https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=683509d4f2f8f0a78388512626e58ff0

Reply:
I was going to file a bug report and then thought about how you'd load the bytes *without* memcpy and, ugh, I guess it's a Hard Problem. *However,* you could still use the same *implementation* (with `reverse()`) rather than using byte swapping, and the compiler optimizes it successfully. I think that'd be easier to understand at a quick glance.